tests/framework/statistics/consumer.py

```diff
@@ -0,0 +1,306 @@
+                 consume_stats,
+                 custom={}):
+        """Initialize a consumer."""
+        self._results = {}  # Aggregated results.
+        self._measurements_defs = defaultdict(MeasurementDef)
+        self._statistics_defs = defaultdict(StatisticDef)
+        self._consume_stats = consume_stats
+        self._custom = custom
+        self._statistics = {}
+        self._iteration = 0
+
+    @abstractmethod
+    def ingest(self, iteration, raw_data):
+        """Abstract method for ingesting the raw result."""
+
+    def consume_result(self, name, value):
+        """Aggregate measurement."""
+        results = self._results.get(name, None)
+        if results is None:
+            self._results[name] = []
+        self._results[name].append(value)
+
+    def consume_custom(self, name, value):
+        """Add custom information."""
+        if not self._custom[self._iteration]:
+            self._custom[self._iteration] = {}
+        self._custom[self._iteration][name] = value
+
+    def consume_stat_def(self, value):
+        """Insert statistics definition."""
+        self._statistics_defs[value.name] = value
+
+    def consume_measurement_def(self, value):
+        """Insert measurement definition."""
+        self._measurements_defs[value.name] = value
+
+    def process(self) -> Dict[str, dict]:
+        """Generate statistics as a dictionary."""
+        # Validate that all statistics have results data.
+        assert len(self._results) is len(self._statistics_defs)
+        for m_name in self._results:
+            assert m_name in self._statistics_defs
+
+        # Validate that all statistics have corresponding measurements data.
+        for stat in self._statistics_defs.values():
+            assert stat.measurement_name in self._measurements_defs
+
+        # Generate consumer stats.
+        for stat in self._statistics_defs.values():
+            m_name = stat.measurement_name
+            stat_unit = self._measurements_defs[m_name].unit
+            self._statistics.setdefault(stat.name, {})[self.UNIT_KEY] \
+                = stat_unit
+
+            # We can either consume directly statistics, or compute them based
+            # on measurements.
+            if self._consume_stats:
+                self._statistics[stat.name][self.RESULT_KEY] = \
+                    stat.apply(self._results[stat.name])
+            else:
+                self._statistics[stat.name][self.RESULT_KEY] = \
+                    stat.apply(self._results[m_name])
+
+            # Check pass criteria.
+            if callable(stat.check) and stat.check.target:
+                res = self._statistics[stat.name][self.RESULT_KEY]
+                assert stat.check(res), "%r criteria for %r failed." \
+                                        " Target: %r vs actual: %r." \
+                                        % (stat.check.name, stat.name,
+                                           str(stat.check.target), str(res))
```

Comment:
I see this as an oversimplification because to have a general ready to use framework we must answer to the following questions:
1. Who will push the data to the aggregator? 
2. What format has the data pushed to the aggregator and who process it?
3. What interface has the entity which pushes data to the aggregator?

The `Producer/Consumer` pattern encapsulates the answers for these questions.
1. A raw data producer will push data to the "aggregator" (Consumer).
2. The "aggregator" (Consumer) is responsible for parsing and extracting the needed information from the produced raw data.
3. A `Producer` will produce raw data if an implementation for `produce` abstract method is provided. More specifically, in the case of a `LambdaProducer`, the contract is that it receives a Pyhton function and optionally kwargs for that function which will produce a raw output (stream of bytes, strings, raw bytes etc). The rest is covered by the inner implementation. 

Let me know your thoughts.